src/validation.cpp

```diff
@@ -1506,61 +1506,68 @@ bool DisconnectBlock(const CBlock &block, CValidationState &state,
 
     if (pfClean) *pfClean = false;
 
-    bool fClean = true;
-
     CBlockUndo blockUndo;
     CDiskBlockPos pos = pindex->GetUndoPos();
     if (pos.IsNull()) return error("DisconnectBlock(): no undo data available");
     if (!UndoReadFromDisk(blockUndo, pos, pindex->pprev->GetBlockHash()))
         return error("DisconnectBlock(): failure reading undo data");
 
+    return ApplyBlockUndo(block, state, pindex, view, blockUndo, pfClean);
+}
+
+bool ApplyBlockUndo(const CBlock &block, CValidationState &state,
+                    const CBlockIndex *pindex, CCoinsViewCache &view,
+                    const CBlockUndo &blockUndo, bool *pfClean) {
+    if (pfClean) *pfClean = false;
+    bool fClean = true;
+
     if (blockUndo.vtxundo.size() + 1 != block.vtx.size())
         return error("DisconnectBlock(): block and undo data inconsistent");
 
-    // undo transactions in reverse order
+    // Undo transactions in reverse order.
     for (int i = block.vtx.size() - 1; i >= 0; i--) {
```

Comment:
It must because of the previous check (and no, blockUndo.vtxundo.size() cannot be size_t.max, there wouldn't be enough addressable memory for this to happen).